### PR TITLE
Add verification function with custom EKU

### PR DESCRIPTION
**Pull request by sietseringers**
This adds a new method called `verify_is_valid_cert_with_eku()` to `EndEntityCert`, similar to `verify_is_valid_tls_server_cert()` and `verify_is_valid_tls_client_cert()` but allowing the caller to specify its own EKU.

We intend to use this crate in our upcoming implementation of the mdoc standard (ISO 18013-5), for use in the [EUDI NL wallet](https://github.com/MinBZK/nl-wallet). That standard uses its own EKUs, such as `1.0.18013.5.1.2`. This change would allow us to use this crate to verify certificate chains having such EKUs.

This PR is loosely based on [this older PR](https://github.com/briansmith/webpki/pull/244) before this project was forked.

### CODE REVIEW

File: src/verify_cert.rs
Comment (cpu) [inline src/verify_cert.rs:354]: I think this could be written to return `Self` instead of a `Result`. The `untrusted::Input::from` is infallible. Was there a reason you were preferring `Result` but always returning `Ok`?

File: src/end_entity.rs
Comment (cpu) [inline src/end_entity.rs:90]: There's a little bit of nuance here that I think should be reflected in the parameter name and doc string.

The `KeyPurposeId` argument here is fed down through to [`check_eku`](https://github.com/rustls/webpki/blob/4a39e2b67d4cddf58b0ea16dd821a04ee2240058/src/verify_cert.rs#L371-L374) in `verify_cert.rs` for the actual validation. That function will return `Ok(())` if the certificate has an EKU specified that includes the required EKU, **or has no EKU specified at all**.

That's the reason the `verify_cert::build_chain` argument is named `required_eku_if_present` instead of `required_eku` and is something I think we should match in this function.

Is that existing behaviour acceptable for the use case you had in mind or would you be expecting `verify_is_valid_cert_with_eku` to return an error if the certificate had no EKU at all?

**Review by cpu — COMMENTED** [review]
Hi @sietseringers, thanks for the PR :-)

I had a couple initial comments. It would also be nice to see a unit test added that covers this new validation function if that's something you would be willing to implement.

**Comment by sietseringers** [comment]
Thanks for your comments! I'll answer them, and I'll also be sure to write some unit tests.

File: src/end_entity.rs
Comment (sietseringers) [reply src/end_entity.rs:90]: Right, that's a good point. I noticed that behaviour of `check_eku`, not sure why I didn't think to call the parameter `required_eku_if_present` like it is there.

ISO 18013-5 requires that the EKU is always present. I thought it might be best to keep this PR small and not modify this aspect of this crate, so in my calling code I now use `x509_parser` (which I already used anyway) to actually enforce the EKU being present.

However, if you agree I would prefer to add an additional boolean called `enforce_eku_presence` or something, pass that along to `check_eku`, and then actually enforce the EKU being pesent if that parameter is set to true. What do you think about that? Alternatively, I'll rename `required_eku` to `required_eku_if_present` instead.

File: src/verify_cert.rs
Comment (sietseringers) [reply src/verify_cert.rs:354]: No, no good reason 😅 I'll change it.

File: src/end_entity.rs
Comment (cpu) [reply src/end_entity.rs:90]: > However, if you agree I would prefer to add an additional boolean called enforce_eku_presence or something, pass that along to check_eku, and then actually enforce the EKU being pesent if that parameter is set to true. What do you think about that?

I think exposing some flexibility for requiring an EKU full-stop seems reasonable to me.

Instead of adding an additional bool parameter, what about changing the `ChainOpts.required_eku_if_present` field from `KeyPurposeId` to an exported enum that has two variants:

```rust
pub enum ExtendedKeyUsage {
    Required(KeyPurposeId),
    RequiredIfPresent(KeyPurposeId),
}
```

The existing code in `verify_is_valid_tls_client_cert` and `verify_is_valid_tls_server_cert` could be updated to build `ChainOptions` with `RequiredIfPresent(verify_cert::EKU_CLIENT_AUTH)` and `RequiredIfPresent(verify_cert::EKU_SERVER_AUTH)`.

Your new `verify_is_valid_cert_with_eku` function could take an `ExtendedKeyUsage` instead of `KeyPurposeId` and pass it straight through to the `ChainOptions` field, letting the caller decide how to enforce the EKU.

Does that seem workable to you? It's possible other maintainers will have their own opinions but I think that would be a nicer representation than an additional `bool` arg.

Comment (sietseringers) [reply src/end_entity.rs:90]: That looks good, I'll do that!

File: src/end_entity.rs
Comment (cpu) [inline src/end_entity.rs:88]: I think `verify_is_valid_tls_server_cert` can now be expressed with a call to `verify_is_valid_cert_with_eku` passing `ExtendedKeyUsage::RequiredIfPresent(verify_cert::EKU_SERVER_AUTH)` as the `eku`. The body of both functions is the same otherwise.

Perhaps we should allow `verify_is_valid_cert_with_eku` to take a `crls` arg so that `verify_is_valid_tls_client_cert` could also be expressed with `verify_is_valid_cert_with_eku` ? 

In retrospect I'm not sure why I didn't add a `crls` arg to `verify_is_valid_tls_server_cert` when adding CRL support. I don't think we're planning to add server certificate CRL validation to Rustls but the existing CRL validation code in webpki isn't specific to client certificates. We should probably hold off on changing the signature of `verify_is_valid_tls_server_cert` for now since it will be a breaking change but for your new function it seems like a good idea.

File: src/end_entity.rs
Comment (cpu) [inline src/end_entity.rs:79]: small nit: we should match the RFC language and use "usage" instead of "use". I also think it might be more natural to write the full version and then the acronym rather than the other way around.

```suggestion
    /// specified extended key usage (EKU).
```

File: src/verify_cert.rs
Comment (cpu) [inline src/verify_cert.rs:65]: What do you think about making this `match` a `fn` in an `impl ExtendedKeyUsage`. Maybe something like `fn key_purpose(&self) -> KeyPurposeId`?

That would let you avoid repeating the same match here and on L388.

File: src/verify_cert.rs
Comment (cpu) [inline src/verify_cert.rs:348]: Small nit: I think it might be clearer written this way that this is a purpose that might be asserted as part of an EKU extension?

```suggestion
/// An OID value indicating a Extended Key Usage (EKU) key purpose.
```

File: src/verify_cert.rs
Comment (cpu) [inline src/verify_cert.rs:358]: I think now that this is `const` we could also tidy the `EKU_SERVER_AUTH`, `EKU_CLIENT_AUTH`, and `EKU_OCSP_SIGNING` constants in this file to use `KeyPurposeId::new` instead of manually initializing the struct.

File: tests/custom_ekus.rs
Comment (cpu) [inline tests/custom_ekus.rs:26]: For the test certs you're using in `verify_custom_eku` I think you can drop `ALL_SIGALGS` and just specify `&[&webpki::RSA_PKCS1_2048_8192_SHA256]` here.

**Review by cpu — COMMENTED** [review]
Thanks for the iteration! This is shaping up nicely :-) 

I had another round of small comments but I think the overall shape looks great.

**Comment by sietseringers** [comment]
Thanks for the review! I implemented the suggestion, and added some more tests; I hope the CI is happy now.

**Comment by sietseringers** [comment]
> Thanks for the review! Good suggestions. I implemented them and added some more tests; I hope the CI is happy now.

Well, almost. Looks like the `main` branch also has the problem it indicates now.

**Comment by cpu** [comment]
> I hope the CI is happy now.

Almost! Looks like there's a couple constants you'll need to gate on the alloc feature to avoid them being considered dead code in the no-default-features build.

FWIW I think if you allow GitHub actions on your fork you can iterate on CI in separate branches as needed. Just mentioning in case one of us misses pressing the "Allow workflows to run" button here after an update.

**Comment by cpu** [comment]
> Looks like the main branch also has the problem it indicates now.

Hmmm the errors I see in [this failed task](https://github.com/rustls/webpki/actions/runs/5546312042/jobs/10126529765?pr=119) are specific to `tests/custom_ekus.rs` which is new to this branch.

**Review by cpu — APPROVED** [review]
This looks good to me, thanks for addressing my earlier feedback. 

@ctz Could you give this a pass and see if you have any reservations? 

I think we'll want to squash-merge this in-tree since the commit history has gotten a little bit long from the iteration.

**Comment by djc** [comment]
Ideally we would land this with a clean commit history that has a separate commit introducing the new data type and refactoring needed to pass that around before in another commit adding the new high-level API (and changing existing API to rely on it).

I noticed (reviewing on mobile so somewhat constrained by the number  of available pixels) that the new method has a different way of taking the trust anchors. Is that well-motivated? In the absence of explicit motivation otherwise I'd be inclined to have them receive arguments the same way.

Not for this PR, but I do wonder if we shouldn't change the top level API here to build a verifier object using a builder API which then gets the EndEntityCert as an argument to a verify method, similar to what @jsha is proposing for the client verifier API in rustls (someone please link in the relevant issue/comment) here, potentially taking ownership of an Arc<[TrustAnchor]> or similar.

**Comment by cpu** [comment]
Thanks @djc (_I didn't tag you explicitly to review only because I knew you're on vacation :zany_face:_)

> I noticed (reviewing on mobile so somewhat constrained by the number of available pixels) that the new method has a different way of taking the trust anchors. Is that well-motivated? In the absence of explicit motivation otherwise I'd be inclined to have them receive arguments the same way.

This was motivated  by my feedback to implement `verify_is_valid_tls_server_cert` and `verify_is_valid_tls_client_cert` via a call to `verify_is_valid_cert_with_eku`. I made that suggestion when I thought the bodies of each were identical save for the `eku` but in retrospect the `trust_anchor` arguments in the function signatures were different to have the type system identify the purpose of the anchors. It might be better to undo that change at the cost of a bit of duplication.

> Not for this PR, but I do wonder if we shouldn't change the top level API here to build a verifier object using a builder API which then gets the EndEntityCert as an argument to a verify method, similar to what @jsha is proposing for the client verifier API in rustls (someone please link in the relevant issue/comment) here, potentially taking ownership of an Arc<[TrustAnchor]> or similar.

https://github.com/rustls/rustls/issues/1360

**Comment by sietseringers** [comment]
> I noticed (reviewing on mobile so somewhat constrained by the number of available pixels) that the new method has a different way of taking the trust anchors. Is that well-motivated? In the absence of explicit motivation otherwise I'd be inclined to have them receive arguments the same way.

My motivation was that while other two functions take `TlsClientTrustAnchors` and `TlsServerTrustAnchors` in the case of `verify_is_valid_cert_with_eku()`, we have trust anchors that are not (generally) of either kind. They are just trust anchors, so I made the function accept a slice directly. I would be happy to add a `TrustAnchors` newtype wrapping
 `&[TrustAnchor]`, however; that would indeed be more consistent with the other two functions.

> This was motivated by my feedback to implement `verify_is_valid_tls_server_cert` and `verify_is_valid_tls_client_cert` via a call to `verify_is_valid_cert_with_eku`. I made that suggestion when I thought the bodies of each were identical save for the `eku` but in retrospect the `trust_anchor` arguments in the function signatures were different to have the type system identify the purpose of the anchors. It might be better to undo that change at the cost of a bit of duplication.

Couldn't I do what I suggested above, and keep the function structure as is by wrapping the anchors in the new `TrustAnchors` type when invoking `verify_is_valid_cert_with_eku()`? That seems reasonable to me and keeps the code deduplication that your suggestion resulted into.

> Ideally we would land this with a clean commit history that has a separate commit introducing the new data type and refactoring needed to pass that around before in another commit adding the new high-level API (and changing existing API to rely on it).

I could close this PR, and open a new one with the commit structure you suggest. While we're at this, am I right in that you prefix the commit messages with the module name that had all or at least most of the changes?

**Comment by cpu** [comment]
> Couldn't I do what I suggested above, and keep the function structure as is by wrapping the anchors in the new TrustAnchors type when invoking verify_is_valid_cert_with_eku()? That seems reasonable to me and keeps the code deduplication that your suggestion resulted into.

Sounds reasonable to me :+1: 

> I could close this PR, and open a new one with the commit structure you suggest.

I think you could also rebase your work in this branch and force push an update, but if a new PR is easier for you I don't think anyone would mind. It's slightly nicer IMO to have all the discussion history in one place.

> am I right in that you prefix the commit messages with the module name that had all or at least most of the changes?

I think that's a nice habit but I don't think it's a practice we apply universally.

**Comment by sietseringers** [comment]
Alright, done (with force push). I also rebased onto the current `main`.

File: tests/custom_ekus.rs
Comment (ctz) [inline tests/custom_ekus.rs:39]: This is a good test for the semantics of `Required`/`RequiredIfPresent`, but really we need an example certificate that demonstrates the real-world data you are using.

There is a real risk that, because this crate is made and maintained for just TLS uses, that we could unknowingly break it for your use case in the future. eg, by tightening up the validation done in `EndEntityCert::try_from`.

**Comment by cpu** [comment]
> ci / coverage (pull_request) Failing after 1m

Looks like a transient error on coveralls side. Can be ignored

File: tests/custom_ekus.rs
Comment (cpu) [reply tests/custom_ekus.rs:39]: +1, I think this is a good suggestion.

**Review by cpu — APPROVED** [review]

**Comment by cpu** [comment]
> Expected — Waiting for status to be reported

Apologies, in https://github.com/rustls/webpki/pull/118 I moved around some CI tasks. If you rebase this branch on main the run+expected checks should line up again.

File: src/trust_anchor.rs
Comment (sietseringers) [reply src/trust_anchor.rs]: Alright, done (and I rebased away the commit to keep the history clean).

File: tests/custom_ekus.rs
Comment (sietseringers) [reply tests/custom_ekus.rs]: Done. I took the example certificates from the ISO spec itself.

**Review by cpu — APPROVED** [review]

File: src/trust_anchor.rs
Comment (sietseringers) [reply src/trust_anchor.rs]: I do note that the invocations of `verify_is_valid_cert_with_eku()` in [`verify_is_valid_tls_server_cert()`](https://github.com/rustls/webpki/pull/119/files#diff-0f57780640df786b48d53e572681d45684de566bd22d5c7aa8c9a67234317bf6R130) and [`verify_is_valid_tls_client_cert`](https://github.com/rustls/webpki/pull/119/files#diff-0f57780640df786b48d53e572681d45684de566bd22d5c7aa8c9a67234317bf6R159) now read a little awkwardly, in the sense that the type `NonTlsTrustAnchors` now occurs in their implementation.

Apart from the original name (`TrustAnchors`) I can't really think of another name that solves that. Perhaps `OtherTrustAnchors` or something, but that seems a little vague.

Comment (ctz) [reply src/trust_anchor.rs]: Yes that looks a bit odd. I would have all three functions dispatch to a single private one that does the work, and takes a `trust_anchors: &[TrustAnchor]`

File: src/verify_cert.rs
Comment (djc) [inline src/verify_cert.rs:397]: I feel like it might be nicer to move this comparison into a method on the new enum? That way it could probably abstract better over the difference of the present vs not variants, and we don't need the the key_purpose_id() method which feels like a bit of a layering violation.

**Review by djc — APPROVED** [review]
This is looking good to me. A few minor suggestions.

**Comment by sietseringers** [comment]
@djc Thanks! I implemented your suggestions.

File: src/trust_anchor.rs
Comment (ctz) [reply src/trust_anchor.rs]: Bump on this point. Happy to provide a small patch myself if that is helpful!

File: src/verify_cert.rs
Comment (sietseringers) [reply src/verify_cert.rs]: Oh, seems I misunderstood @djc's suggestion, apologies. I thought the point was to not have to use the `key_purpose_id()` in that particular place, as opposed to it existing at all.

There is one other place in [`build_chain_inner()`](https://github.com/rustls/webpki/blob/92fd1ba94d443c62c036e176ac50b440d1db5f47/src/verify_cert.rs#L63) where it is used. I can use the same comparison method there, and then `key_purpose_id()` can go away. If that is what you meant then I'll do that.

Comment (cpu) [reply src/verify_cert.rs]: > If that is what you meant then I'll do that.

Yeah I think that's a good approach. Thanks!

Comment (sietseringers) [reply src/verify_cert.rs]: Alright, done.

File: src/end_entity.rs
Comment (cpu) [reply src/end_entity.rs]: @sietseringers I think this piece of minor feedback might have been missed.

Comment (cpu) [reply src/end_entity.rs]: Sorry, my mistake. Was looking at an old revision. It was addressed!

File: src/trust_anchor.rs
Comment (cpu) [reply src/trust_anchor.rs]: @sietseringers Did you catch the above? I think this is missing in the latest branch and the last thing holding up a merge :crossed_fingers:

Comment (sietseringers) [reply src/trust_anchor.rs]: Done!

**Review by ctz — APPROVED** [review]

**Review by cpu — APPROVED** [review]
Thanks!

**Comment by sietseringers** [comment]
Great, thanks! 🎉

**Pull request merged**
